contrib/repro_test.bzl

```diff
@@ -0,0 +1,222 @@
+# Copyright 2017 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Test rule for verifying a given container_image target produces a
+   deterministic Docker image.
+
+This rule builds the given container_image twice in isolation to determine if
+the given image is deterministic.
+It does so by mounting the provided Bazel project files into the 'base' image.
+Next it starts two containers and builds the supplied container_image target
+(being 'image') inside each of the containers. The rule then extracts the
+output files for the test image. Finally, both extracted images are compared.
+A test image is considered to be reproducible when both instances of the same
+image have the same digest and ID. On a mismatch of one of those values, the
+test fails and the container_diff tool
+(https://github.com/GoogleContainerTools/container-diff#container-diff)
+produces the summary of their differences.
+
+Args:
+    image: A container_image target relative to the project's root to test
+           for determinism.
+    base: An image target to build and reproduce the 'image 'in.
+          This image must have Bazel and docker installed and available in the PATH.
+    workspace_file: The WORKSPACE file of the project containing the 'image'
+                    target to help detect project's root path.
+    container_diff_args: (optional) Args to the container_diff tool as specified here:
+                         https://github.com/GoogleContainerTools/container-diff#quickstart
+
+Example:
+
+container_repro_test(
+    name = "set_cmd_repro_test",
+    image = "//tests/container:set_cmd",
+    base = "@bazel_latest//image",
+    workspace_file = "//:WORKSPACE",
+    container_diff_args: ["history", "file", "size", "rpm", "apt", "pip", "node"],
+)
+"""
+
+load("@base_images_docker//util:run.bzl", _extract = "extract")
+load("@bazel_skylib//lib:dicts.bzl", "dicts")
+load("//container:container.bzl", _container = "container")
+load("//skylib:filetype.bzl", container_filetype = "container")
+
+def _impl(ctx):
+    """Core implementation of container_repro_test"""
+
+    name = ctx.attr.name
+    workspace_file = ctx.file.workspace_file
+    proj_root = ctx.actions.declare_file(name + "_src_project_root.txt")
+
+    # Record the absolute path to the provided WORKSPACE file and assume it to
```

Comment:
Is the purpose of this to store the directory of WORKSPACE?
Why not use https://docs.bazel.build/versions/master/skylark/lib/File.html#dirname and store it in a var as opposed to in a file?

Reply:
See reply [here](https://github.com/bazelbuild/rules_docker/pull/955#discussion_r301288032)

Reply:
sgtm